src/topology/uniform_space/equicontinuity.lean

```diff
@@ -0,0 +1,341 @@
+lemma uniform_equicontinuous.uniform_continuous {F : ι → β → α} (h : uniform_equicontinuous F)
+  (i : ι) : uniform_continuous (F i) :=
+λ U hU, mem_map.mpr (mem_of_superset (h U hU) $ λ xy hxy, (hxy i))
+
+protected lemma set.uniform_equicontinuous.uniform_continuous_of_mem {H : set $ β → α}
+  (h : H.uniform_equicontinuous) {f : β → α} (hf : f ∈ H) : uniform_continuous f :=
+h.uniform_continuous ⟨f, hf⟩
+
+/-- Taking sub-families preserves equicontinuity at a point. -/
+lemma equicontinuous_at.comp {F : ι → X → α} {x₀ : X} (h : equicontinuous_at F x₀) (u : κ → ι) :
+  equicontinuous_at (F ∘ u) x₀ :=
+λ U hU, (h U hU).mono (λ x H k, H (u k))
+
+protected lemma set.equicontinuous_at.mono {H H' : set $ X → α} {x₀ : X}
+  (h : H.equicontinuous_at x₀) (hH : H' ⊆ H) : H'.equicontinuous_at x₀ :=
+h.comp (inclusion hH)
+
+/-- Taking sub-families preserves equicontinuity. -/
+lemma equicontinuous.comp {F : ι → X → α} (h : equicontinuous F) (u : κ → ι) :
+  equicontinuous (F ∘ u) :=
+λ x, (h x).comp u
+
+protected lemma set.equicontinuous.mono {H H' : set $ X → α}
+  (h : H.equicontinuous) (hH : H' ⊆ H) : H'.equicontinuous :=
+h.comp (inclusion hH)
+
+/-- Taking sub-families preserves uniform equicontinuity. -/
+lemma uniform_equicontinuous.comp {F : ι → β → α} (h : uniform_equicontinuous F) (u : κ → ι) :
+  uniform_equicontinuous (F ∘ u) :=
+λ U hU, (h U hU).mono (λ x H k, H (u k))
+
+protected lemma set.uniform_equicontinuous.mono {H H' : set $ β → α}
+  (h : H.uniform_equicontinuous) (hH : H' ⊆ H) : H'.uniform_equicontinuous :=
+h.comp (inclusion hH)
+
+/-- A family `𝓕 : ι → X → α` is equicontinuous at `x₀` iff `range 𝓕` is equicontinuous at `x₀`,
+i.e the family `coe : range F → X → α` is equicontinuous at `x₀`. -/
+lemma equicontinuous_at_iff_range {F : ι → X → α} {x₀ : X} :
+  equicontinuous_at F x₀ ↔ equicontinuous_at (coe : range F → X → α) x₀ :=
+⟨λ h, by rw ← comp_range_splitting F; exact h.comp _, λ h, h.comp (range_factorization F)⟩
+
+/-- A family `𝓕 : ι → X → α` is equicontinuous iff `range 𝓕` is equicontinuous,
+i.e the family `coe : range F → X → α` is equicontinuous. -/
+lemma equicontinuous_iff_range {F : ι → X → α} :
+  equicontinuous F ↔ equicontinuous (coe : range F → X → α) :=
+forall_congr (λ x₀, equicontinuous_at_iff_range)
+
+/-- A family `𝓕 : ι → β → α` is uniformly equicontinuous iff `range 𝓕` is uniformly equicontinuous,
+i.e the family `coe : range F → β → α` is uniformly equicontinuous. -/
+lemma uniform_equicontinuous_at_iff_range {F : ι → β → α} :
+  uniform_equicontinuous F ↔ uniform_equicontinuous (coe : range F → β → α) :=
+⟨λ h, by rw ← comp_range_splitting F; exact h.comp _, λ h, h.comp (range_factorization F)⟩
+
+section
+
+local attribute [-instance] Pi.topological_space
+local attribute [-instance] Pi.uniform_space
+local attribute [instance] uniform_convergence.topological_space
+local attribute [instance] uniform_convergence.uniform_space
```

Comment:
You have to do this dance at several places in the PR. Do you think using a type synonym instead would be better, or would it introduce other drawbacks?

Reply:
I will ask on Zulip (as I should have done years ago) because that discussion comes up in every PR using this topology :sweat_smile: 
Before this PR, my main point was that we rarely use this topology on bare functions, so my idea was to have the general setup for uniform convergence topology on bare functions to avoid duplication, but that all facts about it should then be specialized to hom types, so it wasn't worth spending too much energy making the general setup user-friendly. But this indeed suggests that we *do* want the setup to be easily usable for bare functions too, so I'm not sure anymore.
Also imho it doesn't cause that much trouble, but this is probably just because I got used to it.